### PR TITLE
Feat/sebulba ippo

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Implement Sebulba architecture with feedforward IPPO on Rware.
## Why?
Integrate Sebulba's architecture due to its effectiveness in scenarios involving non-jitted/non-jax environments.
## How?
Enhance the existing [Cleanba](https://github.com/vwxyzjn/cleanba) code to support marl algorithm and ensure compatibility with Mava's key components, including logger, evaluator, etc.
## Extra
Action Item: 
   - Update the project's README to reflect these changes.
   - Support absolute metric
   - Add model checkpointer

### CODE REVIEW

File: mava/types.py [hunk 1]
```diff
@@ -87,6 +86,18 @@ class PPOTransition(NamedTuple):
     info: Dict
 
 
+class SebulbaTransition(NamedTuple):
```
Comment (sash-a) [inline mava/types.py:89]: We should try keep this to the same transition. If we're going to make a new transition for each system then those types should go in their own files

File: mava/types.py [hunk 2]
```diff
@@ -118,6 +129,13 @@ class LearnerState(NamedTuple):
     timestep: TimeStep
 
 
+class SebulbaLearnerState(NamedTuple):
```
Comment (sash-a) [inline mava/types.py:132]: Same here, can we not use the existing `LearnerState` it will make the mava API much cleaner because then everything will return a `LearnerState`/`RnnLearnerState`

File: mava/utils/jax.py [hunk 3]
```diff
@@ -42,3 +40,26 @@ def merge_leading_dims(x: chex.Array, num_dims: chex.Numeric) -> chex.Array:
 
     new_shape = (np.prod(x.shape[:num_dims]),) + x.shape[num_dims:]
     return x.reshape(new_shape)
+
+
+def convert_data(
```
Comment (sash-a) [inline mava/utils/jax.py:45]: Need a more explicit name
```suggestion
def add_minibatch_dim(
```

File: mava/utils/sebulba_tools.py
Comment (sash-a) [inline mava/utils/sebulba_tools.py:1]: I think just call it `sebulba.py` as we're already in the `utils` folder

File: mava/utils/sebulba_tools.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    os.environ["TF_XLA_FLAGS"] = "--xla_gpu_autotune_level=2 --xla_gpu_deterministic_reductions"
+    os.environ["TF_CUDNN DETERMINISTIC"] = "1"
+
+    os.environ["JAX_USE_PJRT_C_API_ON_TPU"] = ""
+    # Fix weird OOM https://github.com/google/jax/discussions/6332#discussioncomment-1279991
+    os.environ["XLA_PYTHON_CLIENT_MEM_FRACTION"] = "0.6"
+    os.environ["XLA_FLAGS"] = "--xla_cpu_multi_thread_eigen=false " "intra_op_parallelism_threads=1"
+    # Fix CUDNN non-determinisim; https://github.com/google/jax/issues/4823#issuecomment-952835771
+    os.environ["TF_XLA_FLAGS"] = "--xla_gpu_autotune_level=2 --xla_gpu_deterministic_reductions"
+    os.environ["TF_CUDNN DETERMINISTIC"] = "1"
```
Comment (sash-a) [inline mava/utils/sebulba_tools.py:38]: Same thing twice
```suggestion
```

File: mava/utils/sebulba_tools.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import os
+
+
+def configure_computation_environment() -> None:
+    """Configure the computation environment for JAX.
+
+    Note: Must be called before any JAX computation is run.
+    """
+    os.environ["JAX_USE_PJRT_C_API_ON_TPU"] = ""
+
```
Comment (sash-a) [inline mava/utils/sebulba_tools.py:24]: JAX_USE_PJRT_C_API_ON_TPU no longer has an effect (i.e. it always defaults to true).
```suggestion
```

File: mava/utils/sebulba_tools.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    # Fix weird OOM https://github.com/google/jax/discussions/6332#discussioncomment-1279991
+    os.environ["XLA_PYTHON_CLIENT_MEM_FRACTION"] = "0.6"
```
Comment (sash-a) [inline mava/utils/sebulba_tools.py:26]: I don't think we want this, it's fine to use more than 60% of GPU/TPU memory

File: mava/utils/sebulba_tools.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    os.environ["XLA_FLAGS"] = "--xla_cpu_multi_thread_eigen=false " "intra_op_parallelism_threads=1"
+    # Fix CUDNN non-determinisim; https://github.com/google/jax/issues/4823#issuecomment-952835771
+    os.environ["TF_XLA_FLAGS"] = "--xla_gpu_autotune_level=2 --xla_gpu_deterministic_reductions"
+    os.environ["TF_CUDNN DETERMINISTIC"] = "1"
```
Comment (sash-a) [inline mava/utils/sebulba_tools.py:30]: This is making the code more deterministic at the cost of speed, I think we shouldn't add this, but happy to hear thoughts

File: mava/wrappers/robot_warehouse.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+                        add_agent_id=config.system.add_agent_id,
+                    )
+                    for _ in range(num_envs)
+                ]
+            )
+        envs.num_envs = num_envs
+        envs.is_vector_env = True
+        return envs
+
+    return thunk
```
Comment (sash-a) [inline mava/wrappers/robot_warehouse.py:227]: We'd never want to use `SyncVectorEnv` unless debugging and only one of these needs an inner method.

```suggestion
def make_env(
    map_name: str = "rware-tiny-2ag-v1",
    team_reward: bool = True,
    add_agent_id: bool = True,
) -> Callable[[], gym.Env]:
    """Create a function which creates a fully configured environment."""

    def thunk() -> gym.Env:
        env = GymWrapper(
            env=gym.make(map_name),
            team_reward=team_reward,
        )
        if add_agent_id:
            env = AgentIDWrapper(env)

        return env

    return thunk


def make_envs(num_envs: int, config: DictConfig) -> gym.vector.VectorEnv:
    envs = gym.vector.AsyncVectorEnv(
        [
            make_env(
                map_name="rware-tiny-2ag-v1",
                team_reward=config.system.use_team_reward,
                add_agent_id=config.system.add_agent_id,
            )
            for _ in range(num_envs)
        ]
    )

    envs.num_envs = num_envs
    envs.is_vector_env = True
    return envs
```

File: mava/wrappers/robot_warehouse.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, info
+
+    def step(self, action: list) -> Tuple[np.ndarray, float, bool, bool, Dict]:
+        """Step the environment."""
+        obs, reward, terminated, truncated, info = self.env.step(action)
+        obs = np.concatenate([self.agent_ids, obs], axis=1)
+        return obs, reward, terminated, truncated, info
+
+
```
Comment (sash-a) [inline mava/wrappers/robot_warehouse.py:175]: Couple other comments here, it would be nice if we could integrate this with @WiemKhlifi new make function.

Do we need to set 
```python
       envs.num_envs = num_envs
       envs.is_vector_env = True
```
Do we use them? I think `num_envs` is set automatically anyways

File: pyproject.toml [hunk 10]
```diff
@@ -77,4 +77,5 @@ ignore = [
     "W503",  # Line break before binary operator (not compatible with black).
     "B017",  # assertRaises(Exception): or pytest.raises(Exception) should be considered evil.
     "E203",  # black and flake8 disagree on whitespace before ':'.
+    "E402",  # Module level import not at top of file.
```
Comment (sash-a) [inline pyproject.toml:80]: I think just do a comment ignore, we still want this warning in general

File: mava/logger.py [hunk 11]
```diff
@@ -36,10 +36,22 @@ def __call__(
         ...
 
 
-def get_logger_tools(logger: Logger) -> LogFn:  # noqa: CCR001
+class SebulbaLogFn(Protocol):
```
Comment (sash-a) [inline mava/logger.py:39]: I really don't think this is a good idea, can we please use the existing log_fn otherwise there will be a lot of code duplication

**Review by sash-a — CHANGES_REQUESTED** [review]
Thanks for this @OmaymaMahjoub! Haven't looked at the actual sebulba file, but quite a few fundamental changes, keen to hear your thoughts.

**Comment by OmaymaMahjoub** [comment]
Many changes happened to `deveop` since the last update of this PR that will result in a massive merge conflict, therefore we need to create a new branch and new PR for the sebulba ippo!

File: mava/logger.py (hunk 11 above)
Comment (OmaymaMahjoub) [reply mava/logger.py:39]: I tired to address this issue in #977
